helix-core/src/main/java/org/apache/helix/controller/GenericHelixController.java

```diff
@@ -813,23 +817,36 @@ public void onStateChange(String instanceName, List<CurrentState> statesInfo,
 
   @Override
   @PreFetch(enabled = false)
-  public void onCustomizedStateRootChange(String instanceName, NotificationContext changeContext) {
+  public void onCustomizedStateRootChange(String instanceName, List<String> customizedStateTypes,
+      NotificationContext changeContext) {
     logger.info("START: GenericClusterController.onCustomizedStateRootChange()");
-    notifyCaches(changeContext, ChangeType.CUSTOMIZED_STATE_ROOT);
     HelixManager manager = changeContext.getManager();
-    List<String> customizedStateTypes =
-        manager.getHelixDataAccessor().getChildNames(
-        manager.getHelixDataAccessor().keyBuilder().customizedStatesRoot(instanceName));
+    Builder keyBuilder = new Builder(manager.getClusterName());
 
-    for (String customizedState : customizedStateTypes) {
-      try {
-        manager.addCustomizedStateChangeListener(this, instanceName, customizedState);
-        logger.info(
-            manager.getInstanceName() + " added customized state listener for " + instanceName
-                + ", listener: " + this);
-      } catch (Exception e) {
-        logger.error("Fail to add customized state listener for instance: " + instanceName, e);
+    synchronized (_lastSeenCustomizedStateTypes) {
+      Set<String> lastSeenCustomizedStateTypes = _lastSeenCustomizedStateTypes.get();
```

Comment:
If we already lock on _lastSeenCustomizedStateTypes, do we still need to use AtomicReference?

Reply:
I saw a todo here:
 // TODO: remove the synchronization here once we move this update into dataCache.
    synchronized (_lastSeenInstances) {
I'm not sure whether there's any pending task for this that could also apply here, so I kept it.

Reply:
Then please also carry over the TODO part.